File: puzzle.py
import itertools
import math

from cage import Cage
from cell import Cell


def cage_for_cell(cell_coords: (int, int), cages: list[Cage]) -> Cage:
    return next((cage for cage in cages if cell_coords in cage.coords))


class Puzzle:
    def __init__(self, name, cages):
        self.name = name
        self.cages = cages
        self.cells = [[Cell(cage_for_cell((i, j), cages), (i, j)) for j in range(6)] for i in range(6)]

    def print(self):
        print(f"Puzzle: {self.name}")
        print()

        max_len = max(len(str(cage.value)) for cage in self.cages)
        for row in self.cells:
            print(" ".join((f"{cell.cage.value:>{max_len}}{cell.cage.sign}" for cell in row)))
            if self.has_solution():
                print(" ".join(f"{cell.possibilities[0]:>{max_len + 1}}" for cell in row))
            else:
                space = " " * (max_len + 1 - 3)
                def print_value(i: int, cell: Cell):
                    return (space if i % 2 == 1 else "") + (f"{i}" if i in cell.possibilities else ' ')
                print(" ".join(print_value(i, cell) for cell in row for i in range(1, 3)))
                print(" ".join(print_value(i, cell) for cell in row for i in range(3, 5)))
                print(" ".join(print_value(i, cell) for cell in row for i in range(5, 7)))
        print()

    def iter_cells(self, cage=None):
        for row in self.cells:
            for cell in row:
                if cage is None or cell.cage == cage:
                    yield cell

    @staticmethod
    def satisfies_row_col_constraint(combo, cell_coords):
        # Check every pair of cells in the current combination
        for i in range(len(combo)):
            for j in range(i + 1, len(combo)):
                # If they have the same value, check their coordinates
                if combo[i] == combo[j]:
                    row_i, col_i = cell_coords[i]
                    row_j, col_j = cell_coords[j]

                    # If they are in the same row or same column, this combo is invalid
                    if row_i == row_j or col_i == col_j:
                        return False
        return True

    @staticmethod
    def find_restricted_cell_possibilities(cells: list[Cell], target_product: int, func: callable):
        # cell_domains is a list of lists, e.g., [[1,2,3,4,5,6], [1,2,4,5,6], [5]]
        cell_domains = [c.possibilities for c in cells]
        coords = [c.coords for c in cells]

        n = len(cells)
        cell_possibilities = [set() for _ in range(n)]

        # itertools.product(*list) takes the cartesian product of the specific lists
        for combo in itertools.product(*cell_domains):
            if func(combo) == target_product:
                if Puzzle.satisfies_row_col_constraint(combo, coords):
                    for i, val in enumerate(combo):
                        cell_possibilities[i].add(val)

        return [sorted(list(s)) for s in cell_possibilities]

    def reduce_possibilities_from_cages(self):
        # TODO: 6. In a box of size 2, if one number is required for all remaining possibilities, that number works the same as if it were present in the line
        # TODO: 6a. Bigger boxes
        # TODO: 11. handle cases where duplicates are only allowed if they're not in the same row & col
        # TODO: 12. If a number is required for all possible solutions and only has one possible location, it must be in that location.
        for cage in self.cages:
            if cage.sign == '=':
                self.set_possibilities(cage.coords, [[cage.value]])
            elif cage.sign == '+':
                # find all values 1-6 that can sum to the cage value
                # Addition cage size is not limited
                possibilities = self.find_restricted_cell_possibilities(list(self.iter_cells(cage)), cage.value, sum)
                self.set_possibilities(cage.coords, possibilities)
            elif cage.sign == '-':
                # Subtraction can be a - b or b - a but cage size limit is 2
                possibilities = [i for i in range(1, 7) if
                                 any((i - j == cage.value) or (j - i == cage.value) for j in range(1, 7))]
                self.set_possibilities(cage.coords, [possibilities for c in cage.coords])
            elif cage.sign == '*':
                # Multiplication cage size is not limited
                possibilities = self.find_restricted_cell_possibilities(list(self.iter_cells(cage)), cage.value,
                                                                        math.prod)
                self.set_possibilities(cage.coords, possibilities)
            elif cage.sign == '/':
                # Division can be a / b or b / a but cage size limit is 2
                possibilities = [i for i in range(1, 7) if
                                 any((i / j == cage.value) or (j / i == cage.value) for j in range(1, 7))]
                self.set_possibilities(cage.coords, [possibilities for c in cage.coords])

    def set_possibilities(self, coords: list[tuple[int, int]], possibilities: list[list[int]]):
        for coord, possibility in zip(coords, possibilities):
            cell = self.cells[coord[0]][coord[1]]
            # Intersect instead of replacing
            cell.possibilities = [p for p in cell.possibilities if p in possibility]
            if len(cell.possibilities) < 1:
                raise Exception(
                    f"Invalid cell possibilities: {cell.cage.value}{cell.cage.sign} {cell.coords} {possibilities}")

    def resolved_cells(self):
        return len([cell for cell in self.iter_cells() if len(cell.possibilities) == 1])

    def resolved_possibilities(self):
        return (6 * 6 * 6) - sum(len(cell.possibilities) for cell in self.iter_cells())

    def has_solution(self):
        return all(len(cell.possibilities) == 1 for cell in self.iter_cells())

    def reduce_possibilities_from_rows_and_columns(self):
        def reduce_group(group):
            """Reduces possibilities within a specific group (row or column)."""
            # --- 1. Naked Subsets (Singles, Pairs, Triplets, etc.) ---
            for cell in group:
                poss_set = set(cell.possibilities)
                n = len(poss_set)

                if n == 0:
                    continue

                # Find all cells in this group that share the exact same N possibilities
                matches = [c for c in group if set(c.possibilities) == poss_set]

                # Naked subset rule: N cells with the same N possibilities
                if len(matches) == n:
                    for other_cell in group:
                        if other_cell not in matches:
                            self.set_possibilities([other_cell.coords], [
                                [p for p in other_cell.possibilities if p not in poss_set]])

            # --- 2. Hidden Singles ---
            # Flatten all possibilities in the group to count occurrences
            all_possibilities = [p for c in group for p in c.possibilities]

            for p in set(all_possibilities):
                if all_possibilities.count(p) == 1:
                    # Find the unique cell that contains this possibility
                    self.set_possibilities([cell.coords for cell in group if p in cell.possibilities], [[p]])

        # Apply to rows
        for row in self.cells:
            reduce_group(row)

        # Apply to columns
        num_cols = len(self.cells[0]) if self.cells else 0
        for col_idx in range(num_cols):
            column = [row[col_idx] for row in self.cells]
            reduce_group(column)

File: src/puzzle.ts
import { Cage } from "./cage";
import { Cell } from "./cell";

function cartesianProduct<T>(arrays: T[][]): T[][] {
  if (arrays.length === 0) {
    return [[]];
  }
  return arrays.reduce<T[][]>((acc, array) => {
    return acc.flatMap((partial) => array.map((value) => [...partial, value] as T[]));
  }, [[]]);
}

function arraysEqual<T>(a: T[], b: T[]): boolean {
  if (a.length !== b.length) {
    return false;
  }
  return a.every((value, index) => value === b[index]);
}

function setEquality<T>(a: Set<T>, b: Set<T>): boolean {
  if (a.size !== b.size) {
    return false;
  }
  for (const item of a) {
    if (!b.has(item)) {
      return false;
    }
  }
  return true;
}

export class Puzzle {
  name: string;
  cages: Cage[];
  cells: Cell[][];

  constructor(name: string, cages: Cage[]) {
    this.name = name;
    this.cages = cages;
    this.cells = Array.from({ length: 6 }, (_, i) =>
      Array.from({ length: 6 }, (_, j) => new Cell(Puzzle.cageForCell([i, j], cages), [i, j]))
    );
  }

  static cageForCell(cellCoords: [number, number], cages: Cage[]): Cage {
    const [row, col] = cellCoords;
    const cage = cages.find((cageItem) =>
      cageItem.coords.some(([r, c]) => r === row && c === col)
    );
    if (!cage) {
      throw new Error(`No cage found for cell ${row},${col}`);
    }
    return cage;
  }

  print(): void {
    console.log(`Puzzle: ${this.name}`);
    console.log();

    const maxLen = Math.max(...this.cages.map((cage) => cage.value.toString().length));
    for (const row of this.cells) {
      console.log(
        row
          .map((cell) => `${cell.cage.value.toString().padStart(maxLen, " ")}${cell.cage.sign}`)
          .join(" ")
      );
      if (this.hasSolution()) {
        console.log(
          row
            .map((cell) => cell.possibilities[0].toString().padStart(maxLen + 1, " "))
            .join(" ")
        );
      } else {
        const space = " ".repeat(Math.max(0, maxLen + 1 - 3));
        const printValue = (i: number, cell: Cell): string =>
          (i % 2 === 1 ? space : "") + (cell.possibilities.includes(i) ? i.toString() : " ");
        console.log(
          row.flatMap((cell) => [1, 2].map((i) => printValue(i, cell))).join(" ")
        );
        console.log(
          row.flatMap((cell) => [3, 4].map((i) => printValue(i, cell))).join(" ")
        );
        console.log(
          row.flatMap((cell) => [5, 6].map((i) => printValue(i, cell))).join(" ")
        );
      }
    }
    console.log();
  }

  *iterCells(cage?: Cage): Generator<Cell, void, unknown> {
    for (const row of this.cells) {
      for (const cell of row) {
        if (!cage || cell.cage === cage) {
          yield cell;
        }
      }
    }
  }

  static satisfiesRowColConstraint(combo: number[], cellCoords: Array<[number, number]>): boolean {
    for (let i = 0; i < combo.length; i += 1) {
      for (let j = i + 1; j < combo.length; j += 1) {
        if (combo[i] === combo[j]) {
          const [rowI, colI] = cellCoords[i];
          const [rowJ, colJ] = cellCoords[j];
          if (rowI === rowJ || colI === colJ) {
            return false;
          }
        }
      }
    }
    return true;
  }

  static findRestrictedCellPossibilities(
    cells: Cell[],
    targetProduct: number,
    func: (combo: number[]) => number
  ): number[][] {
    const cellDomains = cells.map((cell) => cell.possibilities);
    const coords = cells.map((cell) => cell.coords);

    const cellPossibilities: Array<Set<number>> = Array.from({ length: cells.length }, () => new Set<number>());

    for (const combo of cartesianProduct(cellDomains)) {
      if (func(combo) === targetProduct && Puzzle.satisfiesRowColConstraint(combo, coords)) {
        combo.forEach((value, index) => cellPossibilities[index].add(value));
      }
    }

    return cellPossibilities.map((set) => Array.from(set).sort((a, b) => a - b));
  }

  reducePossibilitiesFromCages(): void {
    for (const cage of this.cages) {
      if (cage.sign === "=") {
        this.setPossibilities(cage.coords, [ [cage.value] ]);
      } else if (cage.sign === "+") {
        const possibilities = Puzzle.findRestrictedCellPossibilities(
          Array.from(this.iterCells(cage)),
          cage.value,
          (combo) => combo.reduce((acc, value) => acc + value, 0)
        );
        this.setPossibilities(cage.coords, possibilities);
      } else if (cage.sign === "-") {
        const possibilities = Array.from({ length: 6 }, (_, index) => index + 1).filter((i) =>
          Array.from({ length: 6 }, (_, index) => index + 1).some((j) => i - j === cage.value || j - i === cage.value)
        );
        this.setPossibilities(cage.coords, cage.coords.map(() => possibilities));
      } else if (cage.sign === "*") {
        const possibilities = Puzzle.findRestrictedCellPossibilities(
          Array.from(this.iterCells(cage)),
          cage.value,
          (combo) => combo.reduce((acc, value) => acc * value, 1)
        );
        this.setPossibilities(cage.coords, possibilities);
      } else if (cage.sign === "/") {
        const possibilities = Array.from({ length: 6 }, (_, index) => index + 1).filter((i) =>
          Array.from({ length: 6 }, (_, index) => index + 1).some((j) => i / j === cage.value || j / i === cage.value)
        );
        this.setPossibilities(cage.coords, cage.coords.map(() => possibilities));
      }
    }
  }

  setPossibilities(coords: Array<[number, number]>, possibilities: number[][]): void {
    for (let index = 0; index < coords.length; index += 1) {
      const [row, col] = coords[index];
      const cell = this.cells[row][col];
      const allowed = possibilities[index];
      cell.possibilities = cell.possibilities.filter((p) => allowed.includes(p));
      if (cell.possibilities.length < 1) {
        throw new Error(
          `Invalid cell possibilities: ${cell.cage.value}${cell.cage.sign} ${cell.coords[0]},${cell.coords[1]} ${JSON.stringify(
            possibilities
          )}`
        );
      }
    }
  }

  resolvedCells(): number {
    return Array.from(this.iterCells()).filter((cell) => cell.possibilities.length === 1).length;
  }

  resolvedPossibilities(): number {
    return 6 * 6 * 6 - Array.from(this.iterCells()).reduce((sum, cell) => sum + cell.possibilities.length, 0);
  }

  hasSolution(): boolean {
    return Array.from(this.iterCells()).every((cell) => cell.possibilities.length === 1);
  }

  reducePossibilitiesFromRowsAndColumns(): void {
    const reduceGroup = (group: Cell[]): void => {
      for (const cell of group) {
        const possSet = new Set(cell.possibilities);
        const n = possSet.size;
        if (n === 0) {
          continue;
        }
        const matches = group.filter((candidate) => setEquality(new Set(candidate.possibilities), possSet));
        if (matches.length === n) {
          for (const otherCell of group) {
            if (!matches.includes(otherCell)) {
              this.setPossibilities([otherCell.coords], [otherCell.possibilities.filter((p) => !possSet.has(p))]);
            }
          }
        }
      }

      const allPossibilities = group.flatMap((cell) => cell.possibilities);
      const uniquePossibilities = Array.from(new Set(allPossibilities));
      for (const value of uniquePossibilities) {
        const occurrences = allPossibilities.filter((candidate) => candidate === value).length;
        if (occurrences === 1) {
          const coords = group.filter((cell) => cell.possibilities.includes(value)).map((cell) => cell.coords);
          this.setPossibilities(coords, coords.map(() => [value]));
        }
      }
    };

    for (const row of this.cells) {
      reduceGroup(row);
    }

    const numCols = this.cells.length > 0 ? this.cells[0].length : 0;
    for (let colIdx = 0; colIdx < numCols; colIdx += 1) {
      const column = this.cells.map((row) => row[colIdx]);
      reduceGroup(column);
    }
  }
}
